cpp/src/arrow/engine/substrait/util.h

```diff
@@ -0,0 +1,40 @@
+// Licensed to the Apache Software Foundation (ASF) under one
+// or more contributor license agreements. See the NOTICE file
+// distributed with this work for additional information
+// regarding copyright ownership. The ASF licenses this file
+// to you under the Apache License, Version 2.0 (the
+// "License"); you may not use this file except in compliance
+// with the License. You may obtain a copy of the License at
+//
+// http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing,
+// software distributed under the License is distributed on an
+// "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+// KIND, either express or implied. See the License for the
+// specific language governing permissions and limitations
+// under the License.
+
+#pragma once
+
+#include <memory>
+#include "arrow/engine/substrait/api.h"
+#include "arrow/util/iterator.h"
+#include "arrow/util/optional.h"
+
+namespace arrow {
+
+namespace engine {
+
+/// \brief Retrieve a RecordBatchReader from a Substrait plan.
+ARROW_ENGINE_EXPORT Result<std::shared_ptr<RecordBatchReader>> ExecuteSerializedPlan(
+    std::shared_ptr<Buffer> substrait_buffer);
```

Comment:
This should be `const Buffer&`.  That will require some changes to `SubstraitExecutor` but the way it is written right now we are keeping the plan alive for the entire time we execute it and that is incorrect.  We only need the `substrait_buffer` to create the `ExecPlan` (happens in `SubstraitExecutor::Init`).  Once that is done we should allow this buffer to be freed (it's not likely to be a huge chunk of RAM but it seems like good hygiene and the API will be clearer to the user).

Reply:
@westonpace I updated the code. Does it satisfy the requirement?